Marlin/src/module/tool_change.cpp

```diff
@@ -494,225 +477,215 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
-            do_blocking_move_to_z(destination[Z_AXIS], planner.settings.max_feedrate_mm_s[Z_AXIS]);
-          }
+          planner.buffer_line(current_position, feedrate_mm_s, active_extruder);
         #endif
-      } // (tmp_extruder != active_extruder)
+        planner.synchronize();
+      }
 
-      planner.synchronize();
+      #if HOTENDS > 1
+        const float xdiff = hotend_offset[X_AXIS][tmp_extruder] - hotend_offset[X_AXIS][active_extruder],
+                    ydiff = hotend_offset[Y_AXIS][tmp_extruder] - hotend_offset[Y_AXIS][active_extruder],
+                    zdiff = hotend_offset[Z_AXIS][tmp_extruder] - hotend_offset[Z_AXIS][active_extruder];
+      #else
+        const float xdiff = 0, ydiff = 0, zdiff = 0;
+      #endif
 
-      #if ENABLED(EXT_SOLENOID) && DISABLED(PARKING_EXTRUDER)
-        disable_all_solenoids();
-        enable_solenoid_on_active_extruder();
+      #if ENABLED(DUAL_X_CARRIAGE)
+        dualx_tool_change(tmp_extruder, no_move);
+      #elif ENABLED(PARKING_EXTRUDER) // Dual Parking extruder
+        parking_extruder_tool_change(tmp_extruder, no_move);
+      #elif ENABLED(SWITCHING_TOOLHEAD) // Switching Toolhead
+        switching_toolhead_tool_change(tmp_extruder, fr_mm_s, no_move);
+      #elif ENABLED(SWITCHING_NOZZLE)
+        // Always raise by a configured distance to avoid workpiece
+        current_position[Z_AXIS] += MAX(-zdiff, 0.0) + toolchange_settings.z_raise;
+        #if HAS_SOFTWARE_ENDSTOPS
+          NOMORE(current_position[Z_AXIS], soft_endstop_max[Z_AXIS]);
+          update_software_endstops(Z_AXIS);
+        #endif
+        if (!no_move)planner.buffer_line(current_position, planner.settings.max_feedrate_mm_s[Z_AXIS], active_extruder);
+        move_nozzle_servo(tmp_extruder);
       #endif
 
-      feedrate_mm_s = old_feedrate_mm_s;
+      #if ENABLED(DEBUG_LEVELING_FEATURE)
+        if (DEBUGGING(LEVELING)) {
+          SERIAL_ECHOPAIR("Offset Tool XY by { ", xdiff);
+          SERIAL_ECHOPAIR(", ", ydiff);
+          SERIAL_ECHOPAIR(", ", zdiff);
+          SERIAL_ECHOLNPGM(" }");
+        }
+      #endif
 
-      #if HAS_SOFTWARE_ENDSTOPS && ENABLED(DUAL_X_CARRIAGE)
+      // The newly-selected extruder XY is actually at...
+      #if DISABLED(DUAL_X_CARRIAGE)
+        current_position[X_AXIS] += xdiff;
+      #endif
+      current_position[Y_AXIS] += ydiff;
+      current_position[Z_AXIS] += zdiff;
+
+      #if HAS_SOFTWARE_ENDSTOPS
+        NOMORE(current_position[X_AXIS], soft_endstop_max[X_AXIS]);
+        NOLESS(current_position[X_AXIS], soft_endstop_min[X_AXIS]);
+        NOMORE(current_position[Y_AXIS], soft_endstop_max[Y_AXIS]);
+        NOLESS(current_position[Y_AXIS], soft_endstop_min[Y_AXIS]);
+        NOMORE(current_position[Z_AXIS], soft_endstop_max[Z_AXIS]);
+        NOLESS(current_position[Z_AXIS], soft_endstop_min[Z_AXIS]);
         update_software_endstops(X_AXIS);
+        update_software_endstops(Y_AXIS);
+        update_software_endstops(Z_AXIS);
```

Comment:
Same deal here. Nothing in a tool-change, outside of Dual X changing `active_extruder`, causes the software endstops to change, and if they did you'd change them first and then constrain.

The bigger logical error here is that you're altering the `current_position` without making a move. Not a good idea.

Reply:
This is the same condition that is in use for no_move setting tool offsets prior to this change. See current 543-545. I can go back and add a planner.bufferline(current_position) on all of these after settings the offsets if that would be seen as a safer update.